gcs/src/gcs.cpp

```diff
@@ -1762,47 +1762,46 @@ long gcs_destroy (gcs_conn_t *conn)
             if (GCS_CONN_CLOSED > conn->state)
                 gu_error ("Attempt to call gcs_destroy() before gcs_close(): "
                           "state = %d", conn->state);
-
-            gu_cond_destroy (&tmp_cond);
-
-            return -EBADFD;
+            err = -EBADFD;
         }
 
-        /* this should cancel all recv calls */
-        gu_fifo_destroy (conn->recv_q);
-
         gcs_shift_state (conn, GCS_CONN_DESTROYED);
         /* we must unlock the mutex here to allow unfortunate threads
          * to acquire the lock and give up gracefully */
     }
     else {
         gcs_sm_leave (conn->sm);
-        gu_cond_destroy (&tmp_cond);
         err = -EBADFD;
-        return err;
     }
 
     gu_cond_destroy (&tmp_cond);
     gcs_sm_destroy (conn->sm);
 
-    if ((err = gcs_fifo_lite_destroy (conn->repl_q))) {
-        gu_debug ("Error destroying repl FIFO: %d (%s)", err, strerror(-err));
-        return err;
+    /* this should cancel all recv calls */
+    gu_fifo_destroy (conn->recv_q);
+
+    if ((ret = gcs_fifo_lite_destroy (conn->repl_q)))
+    {
+        gu_debug ("Error destroying repl FIFO: %ld (%s)", ret, strerror(-ret));
     }
 
-    if ((err = gcs_core_destroy (conn->core))) {
-        gu_debug ("Error destroying core: %d (%s)", err, strerror(-err));
-        return err;
+    if ((ret = gcs_core_close(conn->core)))
+    {
+        gu_debug ("Failed to close GCS: error: %ld (%s)",-ret, strerror(-ret));
     }
 
+    if ((ret = gcs_core_destroy (conn->core)))
```

Comment:
If we fail here, err is 0, so the function will return 0. Previously it returned error code.

Reply:
Fixed it by capturing the return values from the cleanup functions and assigning it to `err`.

Reply:
As we discussed on slack:
1. Please remove ret variable as only err is enough
2. Please add the comment describing why we do not return after 1st failure and it is still worth calling gcs_core_destroy() even if gcs_core_close() failed.

Reply:
Done